docs/RFCS/20200401_alter_column_type.md

```diff
@@ -0,0 +1,189 @@
+- This includes the column swap and index swap processes defined in A and B, the steps are explained in less detail here.
+1. Create a new non-public computed column c' using c::t'
+2. Make a copy of the primary key pk and replace each reference to c with c'
+    - pk' must be encoded as a primary index, cannot switch secondary index to primary index
+3. Each index i that references pk, create an index i' that references pk'
+4. Enqueue primary index swap with pk and pk', and index swaps with i and i'. 
+    - Includes enqueuing primary key drop mutation for the old primary key.
+5. Enqueue column swap mutation. 
+6. Enqueue drops for c' and old indexes.
+ 
+## Constraints
+During the addition of the new column, we may also have to rewrite constraints. Furthermore, while the new column exists and the old column has not been dropped yet, we have to validate constraints on both the new and the old column.
+1. On creation of new column, copy constraints and create constraint for new column. Each comparison to the column has to be casted to the new type. Comparisons that have the column inside a nested expression must also be converted.
+    - For example, if (x INT) has check (x + 5 > 10), after float conversion, the constraint would be (x + 5::float > 10::float).
+2. Check if constraint is valid (casts and comparisons are valid), if the constraint is invalid, fail the schema change and rollback.
+3. After the column swap, drop the constraints of the old column when the old column is dropped.
+4. Rename the constraints during/after the swap since we can't have duplicate constraint names if we want to keep the constraint name the same.
+
+## Note about USING EXPRESSION
+When an expression is provided to alter the column type, Postgres does not apply the expression to the column's default expression, constraints and indexes.
+
+    
+## Failure during schema change
+When converting column c (type t) to c' (type t'). During the creation of c' as a computed column, if a value x in c cannot be converted to type t' using the conversion function, the column creation fails and we stop the schema change. This state is easy to roll back from as we simply drop the new column c'. This case happens before the swap mutation.
+
+For example, if we try converting a column from string -> int, and we run into a value 'hello', then the computed column cannot be created and a rollback will happen (dropping the column).
+This is similar to any other case of a schema change running into a runtime error.
+
+If the swap mutation has happened - the type of the column is effectively changed, the only thing left to do is drop the old column.
+    
+# Index backfill method
+- **This method is contingent on changing the index backfiller to support column mutations which is currently blocked due an issue with supporting interleaved tables**
+    - Read more in the [Github Issue](https://github.com/cockroachdb/cockroach/issues/47989)
+- Swap columns using primary index swap.
```

Comment:
I'm not sure if this and below sections are needed here. The main idea is that we would use index backfills rather than column backfills, but the logical set of steps is the same. It would be more instructive to just link to the issue and then just talk here about how that work would impact this feature.

Reply:
I removed the steps but left the bit about the differences between column and index backfiller. Also left in the bit about rollbacks since it's slightly different.